File: java/Stack/LargestRectangleHistogram.java
package java.Stack;
import java.util.Stack;


public class LargestRectangleHistogram {
    public int largestRectangleArea(int[] heights) {
        int n = heights.length;
        int maxArea = 0;
        Stack<Integer> stack = new Stack<>();
        
        for (int i = 0; i <= n; i++) {
            // For i == n, we use a height of 0 to force remaining calculations
            int currentHeight = (i == n) ? 0 : heights[i];
            
            // While current bar is shorter than the bar at stack top
            while (!stack.isEmpty() && currentHeight < heights[stack.peek()]) {
                // Pop the bar from stack and calculate area with that bar as height
                int height = heights[stack.pop()];
                
                // Width extends from the previous smaller bar to current bar
                int width = stack.isEmpty() ? i : i - stack.peek() - 1;
                
                maxArea = Math.max(maxArea, height * width);
            }
            stack.push(i);
        }
        
        return maxArea;
    }
}

File: python/Stack/LargestRectangleHistogram.py

"""
84. Largest Rectangle in Histogram
Difficulty: Hard
Topics: Array, Stack, Monotonic Stack
"""

def largestRectangleArea(heights: list[int]) -> int:
    stack = []  # Stores pairs: (index, height)
    max_area = 0
    
    # Iterate through all bars
    for i, h in enumerate(heights):
        start = i
        # While current bar is shorter than the bar at stack top
        while stack and stack[-1][1] > h:
            # Pop the taller bar and calculate area
            index, height = stack.pop()
            # Width extends from popped index to current index
            max_area = max(max_area, height * (i - index))
            # Current bar can extend backward to the popped bar's index
            start = index
        stack.append((start, h))
    
    # Process remaining bars in stack
    # These bars can extend to the end of the histogram
    for i, h in stack:
        max_area = max(max_area, h * (len(heights) - i))
    
    return max_area